include/aws/common/priority_queue.h

```diff
@@ -0,0 +1,88 @@
+#ifndef AWS_PRIORITY_QUEUE_H
+#define AWS_PRIORITY_QUEUE_H
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/common.h>
+#include <aws/common/array_list.h>
+
+struct aws_priority_queue {
+    int(*compare)(void *a, void *b);
+    struct aws_array_list container;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a priority queue struct for use. This mode will grow memory automatically (exponential model)
+     * Default size is the inital size of the queue
+     * item_size is the size of each element in bytes. Mixing items types is not supported by this API.
+     * compare() is the function that will be used to determine priority.
+     * If compare returns a positive integer, a is of higher priority than b.
+     */
+    AWS_COMMON_API int aws_priority_queue_dynamic_init(struct aws_priority_queue *queue,
+            struct aws_allocator *alloc, size_t default_size, size_t item_size, int(*compare)(void *a, void *b));
```

Comment:
compare may be better as a typedef? I don't feel particularly strongly about it though.

Reply:
I will typedef it and document its semantics as @bdonlan mentioned.